### PR TITLE
entity property default

### PR DESCRIPTION
**Pull request by PawelSuwinski**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        | 

Adds generating of entity property default value based on `#ORM\Column` or `#ApiProperty` attribute as doc generator does.

For example:

```schema.yaml
# config/schema.yaml
types:
  Book:
    properties:
      condition:
        range: "https://schema.org/OfferItemCondition"
        attributes: 
          ORM\Column: 
            options: 
              default: "https://schema.org/NewCondition"
```
gives:
```php
// App\Entity\Book
class Book
(...)
   #[Assert\NotNull]
   #[ORM\Column(options: ['default' => 'https://schema.org/NewCondition'])]
    private $condition = 'https://schema.org/NewCondition';
```

For a background. API property metadata factory take into account default value  based on `#ORM\Column`  attribute or  `#ApiProperty` and generates appropriate doc with default keyword. But if this property is send back as not set or it is is out of used normalization group default value will never be persisted as expected (`null` in database or exception on not null requirement). Doctrine attribute `#ORM\Column: { options: { default: (...) } }` is not enough because doctrine can't do this (https://www.doctrine-project.org/projects/doctrine-orm/en/latest/reference/faq.html#how-can-i-add-default-values-to-a-column). Default value should be explicitly set in entity class during property declaration but schema-generator does not generate this defaults. This PR fix it.

### CODE REVIEW

**Comment by PawelSuwinski** [comment]
> I don't think it is the right fix, it is too much magical and very specific to API Platform / Doctrine. What do you think of having a `defaultValue` parameter in the property configuration instead?

I wanted to change the least code possible to get working consistent solution, but I get the point. 
So here it is an implementation with proposed `defaultValue` property configuration.

For example:
```yaml
# config/schema.yaml
types:
  BookFormatType: ~
  Book:
    properties:
      isbn: ~
      bookFormat:
        range: "BookFormatType"
        defaultValue: "https://schema.org/Hardcover"
```

gives:
```php
(...)
    /**
     * The format of the book.
     *
     * @see https://schema.org/bookFormat
     */
    #[ORM\Column(nullable: true, options: ['default' => 'https://schema.org/Hardcover'])]
    #[ApiProperty(types: ['https://schema.org/bookFormat'], default: 'https://schema.org/Hardcover')]
    #[Assert\Choice(callback: [BookFormatType::class, 'toArray'])]
    private ?string $bookFormat = 'https://schema.org/Hardcover';
```

**Comment by alanpoulain** [comment]
Yes, I think it would be great like this!
I'm not convinced about adding the default value to `ORM\Column` and `ApiProperty` attributes though.

**Comment by PawelSuwinski** [comment]
>  I'm not convinced about adding the default value to `ORM\Column` and `ApiProperty` attributes though.

It seemed natural as was suggested by DoctrineOrmPropertyMetadaFactory that already use it to set default in generated api doc:
https://github.com/api-platform/core/blob/a16cb35ec3619f2d63471318468eb437515a7f90/src/Doctrine/Orm/Metadata/Property/DoctrineOrmPropertyMetadataFactory.php#L77

But I don't insist on that. Let me know if I should withdraw this part from PR.

**Comment by PawelSuwinski** [comment]
Done.

File: src/Model/Property.php [hunk 1]
```diff
@@ -127,6 +129,40 @@ public function addRemoverAnnotation(string $annotation): self
...
+    {
+        return $this->isTypeHintedAsCollection()
+                || $this->typeHint && $this->typeHint !== \gettype($this->defaultValue)
+                    // PHP7 workaround for `get_debug_type()` (e.g., int case).
+                    && $this->type !== \gettype($this->defaultValue)
+                || $this->isArray && !\is_array($this->defaultValue)
+                || !$this->isArray && \is_array($this->defaultValue)
+                || $this->isEnum && $this->isArray
+                    // Only simple array is valid for enum.
+                    && array_filter($this->defaultValue, fn ($v) => !\is_scalar($v))
```
Comment (alanpoulain) [inline src/Model/Property.php:145]: Is this really necessary to validate the default value? I think it's the user responsibility to use the correct default value.
If there is a mistake, PHP will not like it when the code is executed anyway.

File: src/Model/Property.php [hunk 2]
```diff
@@ -127,6 +129,40 @@ public function addRemoverAnnotation(string $annotation): self
...
+     * @param mixed $defaultValue
+     */
+    public function setDefaultValue($defaultValue): self
+    {
+        $value = \is_array($defaultValue) ? $defaultValue : [$defaultValue];
+        array_walk_recursive($value, function ($item) {
+            if (!\is_scalar($item)) {
+                throw new \InvalidArgumentException('Only constants allowed as default value.');
+            }
+        });
```
Comment (alanpoulain) [inline src/Model/Property.php:160]: See my previous comment, is this validation really necessary?

**Comment by PawelSuwinski** [comment]
OK, we can take less strict approach moving responsibility to user. So here it is an implementation without validation. 

I looks like some commits added in meanwhile doesn't pass the phpstan build check (?), so my PR with merged those changes doesn't either :/.

**Comment by alanpoulain** [comment]
Thank you @PawelSuwinski!

**Pull request merged**